File: packages/crypto/src/envelope.ts
/**
 * Boot key envelope: delivery of one environment DEK to one boot process.
 *
 * The server generates a fresh X25519 key pair per approval, derives a wrapping key
 * with HKDF-SHA256 over the raw shared secret, and encrypts the DEK with AES-256-GCM
 * using the same info string as the AAD.
 */

import { bootEnvelopeInfo } from "./aad.ts";
import { aesGcmDecrypt, aesGcmEncrypt, AES_NONCE_LENGTH } from "./aesgcm.ts";
import { b64uDecode, b64uEncode, randomBytes, utf8Encode, type Bytes } from "./encoding.ts";
import { keyFingerprint } from "./fingerprint.ts";
import {
  generateX25519PrivateKey,
  importX25519PrivateKey,
  importX25519PublicKey,
  x25519PublicKeyFromPrivate,
  KEY_LENGTH,
} from "./keys.ts";

export const ENVELOPE_SALT_LENGTH = 32;

/** The `keyEnvelope` object as it appears in a `boot.approved` frame. */
export interface BootEnvelope {
  readonly serverPublicKey: string;
  readonly salt: string;
  readonly nonce: string;
  readonly ciphertext: string;
}

export interface DeriveEnvelopeWrapKeyInput {
  readonly privateKey: Bytes;
  readonly peerPublicKey: Bytes;
  readonly salt: Bytes;
  readonly info: string;
}

/**
 * X25519 then HKDF-SHA256 to a 256-bit wrapping key.
 *
 * An all-zero shared secret means the peer sent a low-order point. Reject it.
 */
export async function deriveEnvelopeWrapKey(input: DeriveEnvelopeWrapKeyInput): Promise<Bytes> {
  const privateKey = await importX25519PrivateKey(input.privateKey);
  const peerPublicKey = await importX25519PublicKey(input.peerPublicKey);
  const shared = new Uint8Array(
    await globalThis.crypto.subtle.deriveBits(
      { name: "X25519", public: peerPublicKey },
      privateKey,
      KEY_LENGTH * 8,
    ),
  );
  let sharedBits = 0;
  for (const byte of shared) sharedBits |= byte;
  if (sharedBits === 0) {
    throw new Error("X25519 produced an all-zero shared secret");
  }
  const hkdfKey = await globalThis.crypto.subtle.importKey("raw", shared, "HKDF", false, [
    "deriveBits",
  ]);
  const wrapKey = await globalThis.crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt: input.salt, info: utf8Encode(input.info) },
    hkdfKey,
    KEY_LENGTH * 8,
  );
  return new Uint8Array(wrapKey);
}

/** Fixed randomness for a deterministic envelope. Test vectors only; never pass this in production. */
export interface BootEnvelopeMaterial {
  readonly serverPrivateKey: Bytes;
  readonly salt: Bytes;
  readonly nonce: Bytes;
}

export interface CreateBootEnvelopeInput {
  readonly bootId: string;
  readonly environmentId: string;
  readonly environmentKeyVersion: number;
  readonly environmentKey: Bytes;
  readonly clientPublicKey: Bytes;
  readonly material?: BootEnvelopeMaterial;
}

/** The envelope plus the info string it was bound to. */
export interface CreatedBootEnvelope {
  readonly envelope: BootEnvelope;
  readonly info: string;
}

/** Build the `keyEnvelope` for an approved boot. Runs on the server. */
export async function createBootEnvelope(
  input: CreateBootEnvelopeInput,
): Promise<CreatedBootEnvelope> {
  const material = input.material;
  const serverPrivateKey = material?.serverPrivateKey ?? generateX25519PrivateKey();
  const serverPublicKey = await x25519PublicKeyFromPrivate(serverPrivateKey);
  const salt = material?.salt ?? randomBytes(ENVELOPE_SALT_LENGTH);
  const nonce = material?.nonce ?? randomBytes(AES_NONCE_LENGTH);

  const info = bootEnvelopeInfo({
    bootId: input.bootId,
    environmentId: input.environmentId,
    environmentKeyVersion: input.environmentKeyVersion,
    clientPublicKeyFingerprint: await keyFingerprint(input.clientPublicKey),
    serverPublicKeyFingerprint: await keyFingerprint(serverPublicKey),
  });
  const wrapKey = await deriveEnvelopeWrapKey({
    privateKey: serverPrivateKey,
    peerPublicKey: input.clientPublicKey,
    salt,
    info,
  });
  const ciphertext = await aesGcmEncrypt({
    key: wrapKey,
    nonce,
    plaintext: input.environmentKey,
    aad: info,
  });
  return {
    envelope: {
      serverPublicKey: b64uEncode(serverPublicKey),
      salt: b64uEncode(salt),
      nonce: b64uEncode(nonce),
      ciphertext: b64uEncode(ciphertext),
    },
    info,
  };
}

export interface OpenBootEnvelopeInput {
  readonly bootId: string;
  readonly environmentId: string;
  readonly environmentKeyVersion: number;
  readonly clientPrivateKey: Bytes;
  readonly envelope: BootEnvelope;
}

/** Open the envelope with the boot's X25519 private key. Runs on the client. */
export async function openBootEnvelope(input: OpenBootEnvelopeInput): Promise<Bytes> {
  const serverPublicKey = b64uDecode(input.envelope.serverPublicKey);
  const clientPublicKey = await x25519PublicKeyFromPrivate(input.clientPrivateKey);
  const info = bootEnvelopeInfo({
    bootId: input.bootId,
    environmentId: input.environmentId,
    environmentKeyVersion: input.environmentKeyVersion,
    clientPublicKeyFingerprint: await keyFingerprint(clientPublicKey),
    serverPublicKeyFingerprint: await keyFingerprint(serverPublicKey),
  });
  const wrapKey = await deriveEnvelopeWrapKey({
    privateKey: input.clientPrivateKey,
    peerPublicKey: serverPublicKey,
    salt: b64uDecode(input.envelope.salt),
    info,
  });
  return await aesGcmDecrypt({
    key: wrapKey,
    nonce: b64uDecode(input.envelope.nonce),
    ciphertext: b64uDecode(input.envelope.ciphertext),
    aad: info,
  });
}

File: apps/env-client/internal/vaultcrypto/envelope.go
package vaultcrypto

import (
	"crypto/ecdh"
	"crypto/hkdf"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"errors"
	"fmt"
)

// ErrZeroSharedSecret is returned when X25519 produces an all-zero shared
// secret, which means the peer sent a low-order point.
var ErrZeroSharedSecret = errors.New("x25519: all-zero shared secret")

// KeyEnvelope is the wire form of the environment DEK delivery, with every
// b64u field already decoded.
type KeyEnvelope struct {
	ServerPublicKey []byte
	Salt            []byte
	Nonce           []byte
	Ciphertext      []byte
}

// GenerateX25519 returns a fresh X25519 keypair.
func GenerateX25519() (*ecdh.PrivateKey, error) {
	key, err := ecdh.X25519().GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("x25519: %w", err)
	}
	return key, nil
}

// X25519PublicKey parses a raw 32-byte X25519 public key.
func X25519PublicKey(raw []byte) (*ecdh.PublicKey, error) {
	pub, err := ecdh.X25519().NewPublicKey(raw)
	if err != nil {
		return nil, fmt.Errorf("x25519: %w", err)
	}
	return pub, nil
}

// X25519PrivateKey parses a raw 32-byte X25519 private key.
func X25519PrivateKey(raw []byte) (*ecdh.PrivateKey, error) {
	priv, err := ecdh.X25519().NewPrivateKey(raw)
	if err != nil {
		return nil, fmt.Errorf("x25519: %w", err)
	}
	return priv, nil
}

func sharedSecret(priv *ecdh.PrivateKey, pub *ecdh.PublicKey) ([]byte, error) {
	shared, err := priv.ECDH(pub)
	if err != nil {
		return nil, fmt.Errorf("x25519: %w", err)
	}
	zero := make([]byte, len(shared))
	if subtle.ConstantTimeCompare(shared, zero) == 1 {
		Zero(shared)
		return nil, ErrZeroSharedSecret
	}
	return shared, nil
}

// DeriveWrapKey derives the 32-byte AES key that protects the environment DEK.
func DeriveWrapKey(priv *ecdh.PrivateKey, peer *ecdh.PublicKey, salt, info []byte) ([]byte, error) {
	shared, err := sharedSecret(priv, peer)
	if err != nil {
		return nil, err
	}
	defer Zero(shared)
	key, err := hkdf.Key(sha256.New, shared, salt, string(info), KeySize)
	if err != nil {
		return nil, fmt.Errorf("hkdf: %w", err)
	}
	return key, nil
}

// OpenBootEnvelope recovers the environment DEK from an approval envelope.
// info must be the exact bytes returned by BootEnvelopeInfo, built from the
// fingerprints of the two public keys actually used.
func OpenBootEnvelope(clientPriv *ecdh.PrivateKey, envelope KeyEnvelope, info []byte) ([]byte, error) {
	serverPub, err := X25519PublicKey(envelope.ServerPublicKey)
	if err != nil {
		return nil, err
	}
	wrapKey, err := DeriveWrapKey(clientPriv, serverPub, envelope.Salt, info)
	if err != nil {
		return nil, err
	}
	defer Zero(wrapKey)
	dek, err := Open(wrapKey, envelope.Nonce, envelope.Ciphertext, info)
	if err != nil {
		return nil, err
	}
	if len(dek) != KeySize {
		Zero(dek)
		return nil, fmt.Errorf("boot envelope: dek must be %d bytes, got %d", KeySize, len(dek))
	}
	return dek, nil
}

// SealBootEnvelope builds an approval envelope. The client never calls it.
// It exists for tests and for the fake vault server.
func SealBootEnvelope(serverPriv *ecdh.PrivateKey, clientPub *ecdh.PublicKey, salt, nonce, dek, info []byte) (KeyEnvelope, error) {
	wrapKey, err := DeriveWrapKey(serverPriv, clientPub, salt, info)
	if err != nil {
		return KeyEnvelope{}, err
	}
	defer Zero(wrapKey)
	ciphertext, err := Seal(wrapKey, nonce, dek, info)
	if err != nil {
		return KeyEnvelope{}, err
	}
	return KeyEnvelope{
		ServerPublicKey: serverPriv.PublicKey().Bytes(),
		Salt:            salt,
		Nonce:           nonce,
		Ciphertext:      ciphertext,
	}, nil
}
